src/controller/java/AndroidDeviceControllerWrapper.cpp

```diff
@@ -226,3 +279,132 @@ void AndroidDeviceControllerWrapper::OnPairingDeleted(CHIP_ERROR error)
 {
     CallVoidInt(GetJavaEnv(), mJavaObjectRef, "onPairingDeleted", static_cast<jint>(error));
 }
+
+void AndroidDeviceControllerWrapper::OnMessage(chip::System::PacketBufferHandle msg) {}
+
+void AndroidDeviceControllerWrapper::OnStatusChange(void) {}
+
+void AndroidDeviceControllerWrapper::SetDelegate(PersistentStorageResultDelegate * delegate)
+{
+    mStorageResultDelegate = delegate;
+}
+
+void AndroidDeviceControllerWrapper::GetKeyValue(const char * key)
+{
+    jstring keyString   = NULL;
+    jstring valueString = NULL;
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    jclass storageCls   = GetPersistentStorageClass();
+    jmethodID method    = GetJavaEnv()->GetStaticMethodID(storageCls, "getKeyValue", "(Ljava/lang/String;)Ljava/lang/String;");
+
+    GetJavaEnv()->ExceptionClear();
+
+    err = N2J_NewStringUTF(GetJavaEnv(), key, keyString);
+    SuccessOrExit(err);
+
+    valueString = (jstring) GetJavaEnv()->CallStaticObjectMethod(storageCls, method, keyString);
+
+    if (mStorageResultDelegate)
+    {
+        mStorageResultDelegate->OnValue(key, GetJavaEnv()->GetStringUTFChars(valueString, 0));
+    }
+
+exit:
+    GetJavaEnv()->ExceptionClear();
+    GetJavaEnv()->DeleteLocalRef(keyString);
+    GetJavaEnv()->DeleteLocalRef(valueString);
+}
+
+CHIP_ERROR AndroidDeviceControllerWrapper::GetKeyValue(const char * key, char * value, uint16_t & size)
+{
+    jstring keyString   = NULL;
+    jstring valueString = NULL;
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    jclass storageCls   = GetPersistentStorageClass();
+    jmethodID method    = GetJavaEnv()->GetStaticMethodID(storageCls, "getKeyValue", "(Ljava/lang/String;)Ljava/lang/String;");
+
+    GetJavaEnv()->ExceptionClear();
+
+    err = N2J_NewStringUTF(GetJavaEnv(), key, keyString);
+    SuccessOrExit(err);
+
+    valueString = (jstring) GetJavaEnv()->CallStaticObjectMethod(storageCls, method, keyString);
+
+    if (value != nullptr)
+    {
+        if (valueString != NULL)
+        {
+            size = strlcpy(value, GetJavaEnv()->GetStringUTFChars(valueString, 0), size);
+        }
+        else
+        {
+            size = 0;
```

Comment:
OK, so the `size +=1` was making no sense to me (especially in the `valueString == NULL` case), so I went looking at other implementations of `GetKeyValue`.  Comparing this to `CHIPPersistentStorageDelegateBridge::GetKeyValue` I see where the +=1 is coming from, but the variables are backwards here.  The logic there is:

1. If we have no value, then return error.
2. If the outparam is null, set size to the length of our value + 1 (for the null terminator we would need).
3. If we have a value and a non-null outparam, use strlcpy, then make sure to tell the caller about the space we really needed (which includes the null terminator).

Given that, we want to make these changes to match those semantics:
```suggestion
    if (valueString != NULL)
    {
        if (value != nullptr)
        {
            size = strlcpy(value, GetJavaEnv()->GetStringUTFChars(valueString, 0), size);
        }
        else
        {
            size = GetJavaEnv()->GetStringUTFLength(valueString);
```

or so, right?

Reply:
Thanks for pointing that out - swapped the if conditions.